src/security/core/src/dds_security_timed_cb.c

```diff
@@ -0,0 +1,330 @@
+/*
+ * Copyright(c) 2006 to 2019 ADLINK Technology Limited and others
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0, or the Eclipse Distribution License
+ * v. 1.0 which is available at
+ * http://www.eclipse.org/org/documents/edl-v10.php.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR BSD-3-Clause
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <assert.h>
+
+#include "dds/ddsrt/atomics.h"
+#include "dds/ddsrt/heap.h"
+#include "dds/ddsrt/log.h"
+#include "dds/ddsrt/sync.h"
+#include "dds/ddsrt/threads.h"
+#include "dds/ddsrt/time.h"
+
+
+#include "dds/security/core/dds_security_timed_cb.h"
+
+
+struct dds_security_timed_dispatcher_t
+{
+    bool active;
+    void *listener;
+};
+
+struct queue_event_t
+{
+    struct queue_event_t *next;
```

Comment:
Implementing a priority queue with a linked list is "not so efficient". I'd much rather see it implemented using the priority queue in ddsrt (see ``fibheap.c``), just like in ``q_xevent.c`` or the lease management. It'd also eliminate most of the code, adding an event becomes a simple call to ``insert`` and checking/extracting the next event becomes ``ddsrt_fibheap_min``/``ddsrt_fibheap_extract_min``.

You would still have to keep track of the events associated with a dispatcher maintaining the current behaviour of ``dds_security_timed_dispatcher_free``, but you'd eliminate the scanning of the list of events.

Reply:
The reason why it is currently implemented as a linked list seems to be because multiple dispatchers with multiple event can simultaneously coexist on one thread. There are two options: 1. If we replace the list by the heap we either need to keep a dispatcher list with each its own fibheap and iterate through the list for finding which events have timedout. Keeping one heap and multiple dispatchers in the way it is currently done is not really feasible since nodes in the middle of the tree might need to be removed. 2. There is only one dispatcher per thread and all the resources are allocated to that dispatcher and duplicated if more dispatchers are needed.

Would you say it is better to have the linked list as it currently is or to keep a list of dispatcher with each dispatcher having its own fibheap?

Reply:
I'm not sure I follow: right now there is a single priority queue of events (implemented as a sorted double-linked list), with the only unusual thing being that it skips any events associated with disabled dispatchers. (I don't know when/how/if/why that case is important, but I presume it is.)

Switching to a proper priority queue would mean never having events associated with a disabled dispatcher in the queue. Presumably one would let each disabled dispatcher have a list of events or a priority queue, to be added to the event queue when the dispatcher is enabled. Remains the question what to do when disabling a dispatcher. I'd just keep the current doubly-linked list to track events per dispatcher (minus the ordering).

That way the enable/disable operations get a bit more complicated because they need to add/remove the events from the priority queue, but the upside is that adding and processing become simpler.

Reply:
Thanks for the feedback @eboasson. Unfortunately I also can't clarify what is the use case for enabling/disabling dispatchers but I also assume it is important to keep it.

If I understand your idea correctly, we would keep the information about events in two places, the priority queue and the unordered linked list. If the dispatcher is disabled, events remain in the linked list but are removed from the queue. Once the dispatcher is enabled they are added again to the queue from the linked list. Is my understanding correct?

The main issue I have with that approach is that when an event is triggered it will have to be removed both from the priority queue and the linked list. I assume this to be the most common case and handling it will increase complexity. The second thing is that removing elements from the queue by disabling the dispatcher might result in non-minimum elements having to be removed which can be a costly operation for the queue.

I would see as a better alternative to this having a list of dispatchers, each with its own priority queue.

Reply:
> Is my understanding correct?

Yes.

> The main issue I have with that approach is that when an event is triggered it will have to be removed both from the priority queue and the linked list. I assume this to be the most common case and handling it will increase complexity.

True; and I agree one would assume this to be a common case. I understand in principle how this contraption is/can be useful to the security implementation, but I don't know the details of how it will be used. That means any consideration about common cases versus uncommon case is a bit of a guess.

The model I proposed at least has the advantage that the only potentially expensive operations are enabling/disabling a dispatcher, and those seem to be unlikely to be common.

> The second thing is that removing elements from the queue by disabling the dispatcher might result in non-minimum elements having to be removed which can be a costly operation for the queue.

Removing an arbitrary element (in this implementation) means effectively a decrease-key + extract-min, so worst-case would be O(n log n) or thereabouts (all entries in the queue are owned by the dispatcher). That seems okayish.

> I would see as a better alternative to this having a list of dispatchers, each with its own priority queue.

That should work, too. Instead of a list you could also do a priority queue of active dispatchers, with each dispatcher scheduled based on the earliest event for that dispatcher.

Perhaps @MarcelJordense can shed some light on how this thing is intended to be used? It would be nice to have a better idea of that ...

Reply:
Both the authentication and the access_control plugin provide an interface to set a listener. Note that on each plugin a single listener can be set. The listener is used to inform the user of the authentication and access_control, which is the ddsi core, of identity status changes (not implemented) or of certificates that are invoked (expired). The set_listener is called when initializing the plugin and removed when the plugin is unloaded. Thus the timed callback functionality should allow both the authentication and the access_control plugin to set callbacks on the expiry time of certificates and remove the timeout callback when the corresponding certificate is removed when for example a local or remote participant is deleted.

Reply:
@MarcelJordense Thanks for the context clarification. Could you also clarify if the enabling/disabling of dispatcher is used or needed? The current interface allows creating/deleting dispatchers and enabling/disabling. The latter makes the implementation much harder.

Reply:
Yes enabling/disabling of dispatchers is needed. A dispatcher is associated with a listener that is set on either the authentication and access_control plugin. Thus a dispatcher is enabled when the plugin is initialized and disabled when the plugin is removed.